File: scripts/import_monthly_expenses.php
<?php

declare(strict_types=1);

$baseDir = dirname(__DIR__);
$dbPath = $baseDir . '/db/wallos.db';
$expensesDir = $baseDir . '/Monatliche Ausgaben';

if (!file_exists($dbPath)) {
    fwrite(STDERR, "Database not found: {$dbPath}\n");
    exit(1);
}

$requiredFiles = [
    'Allgemein' => $expensesDir . '/Allgemein.html',
    'Joni' => $expensesDir . '/Joni.html',
    'Einnahmen' => $expensesDir . '/Einnahmen.html',
    'Pro_Person' => $expensesDir . '/Pro_Person.html',
    'Personen_Vergleich' => $expensesDir . '/Personen_Vergleich.html',
];

foreach ($requiredFiles as $label => $path) {
    if (!file_exists($path)) {
        fwrite(STDERR, "Missing {$label} file: {$path}\n");
        exit(1);
    }
}

$db = new SQLite3($dbPath);
$db->busyTimeout(5000);

$userId = 1;
$mainCurrencyId = 1;
$defaultPaymentMethodId = 1;
$defaultCategoryId = 1;
$today = (new DateTime('today'))->format('Y-m-d');

function readTableRows(string $filePath): array
{
    $html = file_get_contents($filePath);
    if ($html === false) {
        return [];
    }

    libxml_use_internal_errors(true);
    $doc = new DOMDocument();
    $doc->loadHTML($html);
    libxml_clear_errors();

    $rows = [];
    foreach ($doc->getElementsByTagName('tr') as $tr) {
        $cells = [];
        foreach ($tr->getElementsByTagName('td') as $td) {
            $cells[] = trim(preg_replace('/\s+/', ' ', html_entity_decode($td->textContent ?? '', ENT_QUOTES | ENT_HTML5)));
        }
        if (count($cells) > 0) {
            $rows[] = $cells;
        }
    }

    return $rows;
}

function parseAmount(string $value): float
{
    $clean = str_replace(["\xc2\xa0", '€', 'EUR', ' '], '', $value);
    $clean = str_replace('.', '', $clean);
    $clean = str_replace(',', '.', $clean);
    $clean = preg_replace('/[^0-9\.\-]/', '', $clean ?? '');
    if ($clean === '' || $clean === null) {
        return 0.0;
    }
    return round((float) $clean, 2);
}

function normalizeName(string $name): string
{
    $name = trim($name);
    $name = preg_replace('/\s+/', ' ', $name);
    return $name ?? '';
}

function parsePeopleList(string $peopleCell): array
{
    $raw = array_map('trim', explode(',', $peopleCell));
    $people = [];
    foreach ($raw as $p) {
        if ($p === '' || $p === '#') {
            continue;
        }
        $p = normalizeName($p);
        if ($p !== '') {
            $people[] = $p;
        }
    }
    return array_values(array_unique($people));
}

function getOrCreateHouseholdId(SQLite3 $db, int $userId, string $name): int
{
    $stmt = $db->prepare('SELECT id FROM household WHERE user_id = :userId AND lower(name) = lower(:name) LIMIT 1');
    $stmt->bindValue(':userId', $userId, SQLITE3_INTEGER);
    $stmt->bindValue(':name', $name, SQLITE3_TEXT);
    $result = $stmt->execute();
    $row = $result ? $result->fetchArray(SQLITE3_ASSOC) : false;
    if ($row && isset($row['id'])) {
        return (int) $row['id'];
    }

    $insert = $db->prepare('INSERT INTO household (name, user_id) VALUES (:name, :userId)');
    $insert->bindValue(':name', $name, SQLITE3_TEXT);
    $insert->bindValue(':userId', $userId, SQLITE3_INTEGER);
    $insert->execute();
    return (int) $db->lastInsertRowID();
}

function upsertSubscription(
    SQLite3 $db,
    int $userId,
    int $currencyId,
    string $name,
    float $price,
    int $cycle,
    int $frequency,
    string $notes,
    string $startDate,
    int $payerUserId,
    int $paymentMethodId,
    int $categoryId
): int {
    $select = $db->prepare('SELECT id FROM subscriptions WHERE user_id = :userId AND lower(name) = lower(:name) LIMIT 1');
    $select->bindValue(':userId', $userId, SQLITE3_INTEGER);
    $select->bindValue(':name', $name, SQLITE3_TEXT);
    $result = $select->execute();
    $row = $result ? $result->fetchArray(SQLITE3_ASSOC) : false;

    if ($row && isset($row['id'])) {
        $id = (int) $row['id'];
        $update = $db->prepare(
            'UPDATE subscriptions
             SET price = :price,
                 currency_id = :currencyId,
                 cycle = :cycle,
                 frequency = :frequency,
                 notes = :notes,
                 start_date = :startDate,
                 next_payment = :startDate,
                 payment_method_id = :paymentMethodId,
                 payer_user_id = :payerUserId,
                 category_id = :categoryId,
                 inactive = 0
             WHERE id = :id AND user_id = :userId'
        );
        $update->bindValue(':price', $price, SQLITE3_FLOAT);
        $update->bindValue(':currencyId', $currencyId, SQLITE3_INTEGER);
        $update->bindValue(':cycle', $cycle, SQLITE3_INTEGER);
        $update->bindValue(':frequency', $frequency, SQLITE3_INTEGER);
        $update->bindValue(':notes', $notes, SQLITE3_TEXT);
        $update->bindValue(':startDate', $startDate, SQLITE3_TEXT);
        $update->bindValue(':paymentMethodId', $paymentMethodId, SQLITE3_INTEGER);
        $update->bindValue(':payerUserId', $payerUserId, SQLITE3_INTEGER);
        $update->bindValue(':categoryId', $categoryId, SQLITE3_INTEGER);
        $update->bindValue(':id', $id, SQLITE3_INTEGER);
        $update->bindValue(':userId', $userId, SQLITE3_INTEGER);
        $update->execute();
        return $id;
    }

    $insert = $db->prepare(
        'INSERT INTO subscriptions
        (name, logo, price, currency_id, next_payment, cycle, frequency, notes, payment_method_id, payer_user_id, category_id, notify, inactive, url, notify_days_before, user_id, cancellation_date, replacement_subscription_id, auto_renew, start_date)
         VALUES
        (:name, "", :price, :currencyId, :startDate, :cycle, :frequency, :notes, :paymentMethodId, :payerUserId, :categoryId, 0, 0, "", 3, :userId, NULL, NULL, 1, :startDate)'
    );
    $insert->bindValue(':name', $name, SQLITE3_TEXT);
    $insert->bindValue(':price', $price, SQLITE3_FLOAT);
    $insert->bindValue(':currencyId', $currencyId, SQLITE3_INTEGER);
    $insert->bindValue(':startDate', $startDate, SQLITE3_TEXT);
    $insert->bindValue(':cycle', $cycle, SQLITE3_INTEGER);
    $insert->bindValue(':frequency', $frequency, SQLITE3_INTEGER);
    $insert->bindValue(':notes', $notes, SQLITE3_TEXT);
    $insert->bindValue(':paymentMethodId', $paymentMethodId, SQLITE3_INTEGER);
    $insert->bindValue(':payerUserId', $payerUserId, SQLITE3_INTEGER);
    $insert->bindValue(':categoryId', $categoryId, SQLITE3_INTEGER);
    $insert->bindValue(':userId', $userId, SQLITE3_INTEGER);
    $insert->execute();
    return (int) $db->lastInsertRowID();
}

function setSubscriptionParticipants(SQLite3 $db, int $subscriptionId, array $householdIds, float $totalPrice): void
{
    if (count($householdIds) === 0) {
        return;
    }

    $delete = $db->prepare('DELETE FROM subscription_participants WHERE subscription_id = :subscriptionId');
    $delete->bindValue(':subscriptionId', $subscriptionId, SQLITE3_INTEGER);
    $delete->execute();

    $totalCents = (int) round($totalPrice * 100);
    $count = count($householdIds);
    $base = intdiv($totalCents, $count);
    $remainder = $totalCents % $count;

    $insert = $db->prepare('INSERT INTO subscription_participants (subscription_id, household_id, amount, is_manual) VALUES (:subscriptionId, :householdId, :amount, 0)');

    foreach (array_values($householdIds) as $idx => $householdId) {
        $cents = $base + ($idx < $remainder ? 1 : 0);
        $amount = round($cents / 100, 2);
        $insert->bindValue(':subscriptionId', $subscriptionId, SQLITE3_INTEGER);
        $insert->bindValue(':householdId', $householdId, SQLITE3_INTEGER);
        $insert->bindValue(':amount', $amount, SQLITE3_FLOAT);
        $insert->execute();
    }
}

function mapMonthsToCycle(int $months): ?array
{
    if ($months === 1) {
        return ['cycle' => 3, 'frequency' => 1];
    }
    if ($months === 3) {
        return ['cycle' => 3, 'frequency' => 3];
    }
    if ($months === 6) {
        return ['cycle' => 3, 'frequency' => 6];
    }
    if ($months === 12) {
        return ['cycle' => 4, 'frequency' => 1];
    }
    return null;
}

$db->exec('BEGIN');

$stats = [
    'household_created' => 0,
    'subscriptions_upserted' => 0,
    'income_recurring_upserted' => 0,
    'income_entries_inserted' => 0,
];

$existingHouseholdCount = (int) $db->querySingle("SELECT COUNT(*) FROM household WHERE user_id = {$userId}");

// 1) Allgemein: shared subscriptions with participants.
$allgemeinRows = readTableRows($requiredFiles['Allgemein']);
$subscriptionIdByName = [];
foreach ($allgemeinRows as $cells) {
    $name = normalizeName($cells[0] ?? '');
    if ($name === '' || strtolower($name) === 'produkt') {
        continue;
    }

    $people = parsePeopleList($cells[1] ?? '');
    $monthlyAmount = parseAmount($cells[2] ?? '');
    $yearlyAmount = parseAmount($cells[3] ?? '');
    $note = normalizeName($cells[5] ?? '');

    if ($monthlyAmount <= 0.0 && $yearlyAmount <= 0.0) {
        continue;
    }

    if ($monthlyAmount > 0.0) {
        $price = $monthlyAmount;
        $cycle = 3;
        $frequency = 1;
    } else {
        $price = $yearlyAmount;
        $cycle = 4;
        $frequency = 1;
    }

    $householdIds = [];
    foreach ($people as $personName) {
        $householdId = getOrCreateHouseholdId($db, $userId, $personName);
        $householdIds[] = $householdId;
    }

    if (count($householdIds) === 0) {
        continue;
    }

    $payerUserId = $householdIds[0];
    $subscriptionId = upsertSubscription($db, $userId, $mainCurrencyId, $name, $price, $cycle, $frequency, $note, $today, $payerUserId, $defaultPaymentMethodId, $defaultCategoryId);
    setSubscriptionParticipants($db, $subscriptionId, $householdIds, $price);
    $subscriptionIdByName[mb_strtolower($name)] = $subscriptionId;
    $stats['subscriptions_upserted']++;
}

// 2) Joni: personal costs (single participant Joni), skip aggregate/helper rows.
$joniRows = readTableRows($requiredFiles['Joni']);
foreach ($joniRows as $cells) {
    $name = normalizeName($cells[0] ?? '');
    if ($name === '' || strtolower($name) === 'name') {
        continue;
    }

    $lowerName = mb_strtolower($name);
    if (strpos($lowerName, 'kosten aus allgemein') !== false) {
        continue;
    }

    $person = normalizeName($cells[1] ?? '');
    if ($person === '' || $person !== 'Joni') {
        continue;
    }

    $monthlyAmount = parseAmount($cells[2] ?? '');
    $yearlyAmount = parseAmount($cells[3] ?? '');
    $note = normalizeName($cells[4] ?? '');

    if ($monthlyAmount <= 0.0 && $yearlyAmount <= 0.0) {
        continue;
    }

    if ($monthlyAmount > 0.0) {
        $price = $monthlyAmount;
        $cycle = 3;
        $frequency = 1;
    } else {
        $price = $yearlyAmount;
        $cycle = 4;
        $frequency = 1;
    }

    $joniHouseholdId = getOrCreateHouseholdId($db, $userId, 'Joni');
    $subscriptionId = upsertSubscription($db, $userId, $mainCurrencyId, $name, $price, $cycle, $frequency, $note, $today, $joniHouseholdId, $defaultPaymentMethodId, $defaultCategoryId);
    setSubscriptionParticipants($db, $subscriptionId, [$joniHouseholdId], $price);
    $subscriptionIdByName[mb_strtolower($name)] = $subscriptionId;
    $stats['subscriptions_upserted']++;
}

// 3) Einnahmen: money expected back from people.
//    Use recurring rows for 1/3/6/12 month patterns, otherwise create one-time entry.
$einnahmenRows = readTableRows($requiredFiles['Einnahmen']);
foreach ($einnahmenRows as $cells) {
    $person = normalizeName($cells[0] ?? '');
    if ($person === '' || mb_strtolower($person) === 'name') {
        continue;
    }

    $amount = parseAmount($cells[1] ?? '');
    $product = normalizeName($cells[2] ?? '');
    $months = (int) parseAmount($cells[3] ?? '0');
    $status = normalizeName($cells[4] ?? '');

    if ($amount <= 0.0) {
        continue;
    }

    if (mb_strtolower($status) === 'ignore') {
        continue;
    }

    $householdId = getOrCreateHouseholdId($db, $userId, $person);
    $subscriptionId = null;
    $productKey = mb_strtolower($product);
    if (isset($subscriptionIdByName[$productKey])) {
        $subscriptionId = $subscriptionIdByName[$productKey];
    }

    $note = "Import Einnahmen: {$product}";
    $mappedCycle = mapMonthsToCycle($months);

    if ($mappedCycle !== null) {
        $check = $db->prepare(
            'SELECT id FROM person_income_recurring
             WHERE user_id = :userId
               AND household_id = :householdId
               AND amount = :amount
               AND cycle = :cycle
               AND frequency = :frequency
               AND ifnull(subscription_id, 0) = ifnull(:subscriptionId, 0)
             LIMIT 1'
        );
        $check->bindValue(':userId', $userId, SQLITE3_INTEGER);
        $check->bindValue(':householdId', $householdId, SQLITE3_INTEGER);
        $check->bindValue(':amount', $amount, SQLITE3_FLOAT);
        $check->bindValue(':cycle', $mappedCycle['cycle'], SQLITE3_INTEGER);
        $check->bindValue(':frequency', $mappedCycle['frequency'], SQLITE3_INTEGER);
        $check->bindValue(':subscriptionId', $subscriptionId, SQLITE3_INTEGER);
        $found = $check->execute();
        $foundRow = $found ? $found->fetchArray(SQLITE3_ASSOC) : false;

        if ($foundRow && isset($foundRow['id'])) {
            $update = $db->prepare(
                'UPDATE person_income_recurring
                 SET note = :note, active = 1, start_date = :startDate
                 WHERE id = :id AND user_id = :userId'
            );
            $update->bindValue(':note', $note, SQLITE3_TEXT);
            $update->bindValue(':startDate', $today, SQLITE3_TEXT);
            $update->bindValue(':id', (int) $foundRow['id'], SQLITE3_INTEGER);
            $update->bindValue(':userId', $userId, SQLITE3_INTEGER);
            $update->execute();
        } else {
            $insert = $db->prepare(
                'INSERT INTO person_income_recurring
                (user_id, household_id, amount, currency_id, cycle, frequency, start_date, end_date, subscription_id, note, active)
                 VALUES
                (:userId, :householdId, :amount, :currencyId, :cycle, :frequency, :startDate, NULL, :subscriptionId, :note, 1)'
            );
            $insert->bindValue(':userId', $userId, SQLITE3_INTEGER);
            $insert->bindValue(':householdId', $householdId, SQLITE3_INTEGER);
            $insert->bindValue(':amount', $amount, SQLITE3_FLOAT);
            $insert->bindValue(':currencyId', $mainCurrencyId, SQLITE3_INTEGER);
            $insert->bindValue(':cycle', $mappedCycle['cycle'], SQLITE3_INTEGER);
            $insert->bindValue(':frequency', $mappedCycle['frequency'], SQLITE3_INTEGER);
            $insert->bindValue(':startDate', $today, SQLITE3_TEXT);
            $insert->bindValue(':subscriptionId', $subscriptionId, SQLITE3_INTEGER);
            $insert->bindValue(':note', $note, SQLITE3_TEXT);
            $insert->execute();
        }
        $stats['income_recurring_upserted']++;
    } else {
        $insert = $db->prepare(
            'INSERT INTO person_income_entries
            (user_id, household_id, amount, currency_id, income_date, subscription_id, note)
             VALUES
            (:userId, :householdId, :amount, :currencyId, :incomeDate, :subscriptionId, :note)'
        );
        $insert->bindValue(':userId', $userId, SQLITE3_INTEGER);
        $insert->bindValue(':householdId', $householdId, SQLITE3_INTEGER);
        $insert->bindValue(':amount', $amount, SQLITE3_FLOAT);
        $insert->bindValue(':currencyId', $mainCurrencyId, SQLITE3_INTEGER);
        $insert->bindValue(':incomeDate', $today, SQLITE3_TEXT);
        $insert->bindValue(':subscriptionId', $subscriptionId, SQLITE3_INTEGER);
        $insert->bindValue(':note', $note . " ({$months} Monate)", SQLITE3_TEXT);
        $insert->execute();
        $stats['income_entries_inserted']++;
    }
}

$newHouseholdCount = (int) $db->querySingle("SELECT COUNT(*) FROM household WHERE user_id = {$userId}");
$stats['household_created'] = max(0, $newHouseholdCount - $existingHouseholdCount);

$db->exec('COMMIT');
$db->close();

echo "Import complete.\n";
echo "Household created: {$stats['household_created']}\n";
echo "Subscriptions upserted: {$stats['subscriptions_upserted']}\n";
echo "Recurring incomes upserted: {$stats['income_recurring_upserted']}\n";
echo "One-time incomes inserted: {$stats['income_entries_inserted']}\n";

?>

File: scripts/import_monthly_expenses.py
#!/usr/bin/env python3
import datetime
import html
import re
import sqlite3
from pathlib import Path


BASE_DIR = Path(__file__).resolve().parent.parent
DB_PATH = BASE_DIR / "db" / "wallos.db"
EXPENSES_DIR = BASE_DIR / "Monatliche Ausgaben"

USER_ID = 1
MAIN_CURRENCY_ID = 1
TODAY = datetime.date.today().isoformat()
DEFAULT_PAYMENT_METHOD_ID = 1
DEFAULT_CATEGORY_ID = 1
DEFAULT_PAYER_ID = 1

FILES = {
    "Allgemein": EXPENSES_DIR / "Allgemein.html",
    "Joni": EXPENSES_DIR / "Joni.html",
    "Einnahmen": EXPENSES_DIR / "Einnahmen.html",
    "Pro_Person": EXPENSES_DIR / "Pro_Person.html",
    "Personen_Vergleich": EXPENSES_DIR / "Personen_Vergleich.html",
}


def normalize_text(value: str) -> str:
    return re.sub(r"\s+", " ", value.strip())


def parse_amount(value: str) -> float:
    cleaned = (
        value.replace("\xa0", "")
        .replace("€", "")
        .replace("EUR", "")
        .replace(" ", "")
        .replace(".", "")
        .replace(",", ".")
    )
    cleaned = re.sub(r"[^0-9.\-]", "", cleaned)
    if not cleaned:
        return 0.0
    return round(float(cleaned), 2)


def parse_people(people_cell: str) -> list[str]:
    result: list[str] = []
    for part in people_cell.split(","):
        person = normalize_text(part)
        if person and person != "#":
            result.append(person)
    seen = set()
    unique = []
    for person in result:
        key = person.lower()
        if key not in seen:
            unique.append(person)
            seen.add(key)
    return unique


def read_table_rows(path: Path) -> list[list[str]]:
    raw = path.read_text(encoding="utf-8", errors="ignore")
    rows: list[list[str]] = []
    tr_matches = re.findall(r"<tr[^>]*>(.*?)</tr>", raw, flags=re.IGNORECASE | re.DOTALL)
    for tr_html in tr_matches:
        td_matches = re.findall(r"<td[^>]*>(.*?)</td>", tr_html, flags=re.IGNORECASE | re.DOTALL)
        cells = []
        for td_html in td_matches:
            text = re.sub(r"<[^>]+>", " ", td_html)
            text = normalize_text(html.unescape(text))
            cells.append(text)
        if cells:
            rows.append(cells)
    return rows


def get_or_create_household_id(conn: sqlite3.Connection, name: str) -> int:
    cur = conn.execute(
        "SELECT id FROM household WHERE user_id = ? AND lower(name) = lower(?) LIMIT 1",
        (USER_ID, name),
    )
    row = cur.fetchone()
    if row:
        return int(row[0])

    cur = conn.execute(
        "INSERT INTO household (name, user_id) VALUES (?, ?)",
        (name, USER_ID),
    )
    return int(cur.lastrowid)


def upsert_subscription(
    conn: sqlite3.Connection,
    name: str,
    price: float,
    cycle: int,
    frequency: int,
    notes: str,
    payer_id: int,
) -> int:
    cur = conn.execute(
        "SELECT id FROM subscriptions WHERE user_id = ? AND lower(name) = lower(?) LIMIT 1",
        (USER_ID, name),
    )
    row = cur.fetchone()
    if row:
        sub_id = int(row[0])
        conn.execute(
            """
            UPDATE subscriptions
               SET price = ?,
                   currency_id = ?,
                   cycle = ?,
                   frequency = ?,
                   notes = ?,
                   start_date = ?,
                   next_payment = ?,
                   payment_method_id = ?,
                   payer_user_id = ?,
                   category_id = ?,
                   inactive = 0
             WHERE id = ? AND user_id = ?
            """,
            (
                price,
                MAIN_CURRENCY_ID,
                cycle,
                frequency,
                notes,
                TODAY,
                TODAY,
                DEFAULT_PAYMENT_METHOD_ID,
                payer_id,
                DEFAULT_CATEGORY_ID,
                sub_id,
                USER_ID,
            ),
        )
        return sub_id

    cur = conn.execute(
        """
        INSERT INTO subscriptions
            (name, logo, price, currency_id, next_payment, cycle, frequency, notes,
             payment_method_id, payer_user_id, category_id, notify, inactive, url,
             notify_days_before, user_id, cancellation_date, replacement_subscription_id,
             auto_renew, start_date)
        VALUES
            (?, '', ?, ?, ?, ?, ?, ?, ?, ?, ?, 0, 0, '', 3, ?, NULL, NULL, 1, ?)
        """,
        (
            name,
            price,
            MAIN_CURRENCY_ID,
            TODAY,
            cycle,
            frequency,
            notes,
            DEFAULT_PAYMENT_METHOD_ID,
            payer_id,
            DEFAULT_CATEGORY_ID,
            USER_ID,
            TODAY,
        ),
    )
    return int(cur.lastrowid)


def set_participants(conn: sqlite3.Connection, sub_id: int, household_ids: list[int], total_price: float) -> None:
    if not household_ids:
        return

    conn.execute("DELETE FROM subscription_participants WHERE subscription_id = ?", (sub_id,))

    total_cents = round(total_price * 100)
    count = len(household_ids)
    base = total_cents // count
    remainder = total_cents % count

    for idx, household_id in enumerate(household_ids):
        cents = base + (1 if idx < remainder else 0)
        amount = round(cents / 100, 2)
        conn.execute(
            "INSERT INTO subscription_participants (subscription_id, household_id, amount, is_manual) VALUES (?, ?, ?, 0)",
            (sub_id, household_id, amount),
        )


def map_months(months: int):
    if months == 1:
        return (3, 1)   # monthly
    if months == 3:
        return (3, 3)   # every 3 months
    if months == 6:
        return (3, 6)   # every 6 months
    if months == 12:
        return (4, 1)   # yearly
    return None


def main() -> None:
    for label, p in FILES.items():
        if not p.exists():
            raise FileNotFoundError(f"Missing {label}: {p}")

    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    conn.execute("BEGIN")

    before_households = conn.execute(
        "SELECT COUNT(*) FROM household WHERE user_id = ?",
        (USER_ID,),
    ).fetchone()[0]

    stats = {
        "subscriptions_upserted": 0,
        "income_recurring_upserted": 0,
        "income_entries_inserted": 0,
    }

    sub_id_by_name: dict[str, int] = {}

    # Allgemein import
    for cells in read_table_rows(FILES["Allgemein"]):
        name = normalize_text(cells[0] if len(cells) > 0 else "")
        if not name or name.lower() == "produkt":
            continue

        people = parse_people(cells[1] if len(cells) > 1 else "")
        monthly = parse_amount(cells[2] if len(cells) > 2 else "")
        yearly = parse_amount(cells[3] if len(cells) > 3 else "")
        note = normalize_text(cells[5] if len(cells) > 5 else "")

        if monthly <= 0 and yearly <= 0:
            continue

        if monthly > 0:
            price, cycle, frequency = monthly, 3, 1
        else:
            price, cycle, frequency = yearly, 4, 1

        household_ids = [get_or_create_household_id(conn, p) for p in people]
        if not household_ids:
            continue

        payer_id = household_ids[0]
        sub_id = upsert_subscription(conn, name, price, cycle, frequency, note, payer_id)
        set_participants(conn, sub_id, household_ids, price)
        sub_id_by_name[name.lower()] = sub_id
        stats["subscriptions_upserted"] += 1

    # Joni import
    for cells in read_table_rows(FILES["Joni"]):
        name = normalize_text(cells[0] if len(cells) > 0 else "")
        if not name or name.lower() == "name":
            continue
        if "kosten aus allgemein" in name.lower():
            continue

        person = normalize_text(cells[1] if len(cells) > 1 else "")
        if person != "Joni":
            continue

        monthly = parse_amount(cells[2] if len(cells) > 2 else "")
        yearly = parse_amount(cells[3] if len(cells) > 3 else "")
        note = normalize_text(cells[4] if len(cells) > 4 else "")
        if monthly <= 0 and yearly <= 0:
            continue

        if monthly > 0:
            price, cycle, frequency = monthly, 3, 1
        else:
            price, cycle, frequency = yearly, 4, 1

        joni_id = get_or_create_household_id(conn, "Joni")
        sub_id = upsert_subscription(conn, name, price, cycle, frequency, note, joni_id)
        set_participants(conn, sub_id, [joni_id], price)
        sub_id_by_name[name.lower()] = sub_id
        stats["subscriptions_upserted"] += 1

    # Einnahmen import
    for cells in read_table_rows(FILES["Einnahmen"]):
        person = normalize_text(cells[0] if len(cells) > 0 else "")
        if not person or person.lower() == "name":
            continue

        amount = parse_amount(cells[1] if len(cells) > 1 else "")
        product = normalize_text(cells[2] if len(cells) > 2 else "")
        months = int(parse_amount(cells[3] if len(cells) > 3 else "0"))
        status = normalize_text(cells[4] if len(cells) > 4 else "")
        if amount <= 0:
            continue
        if status.lower() == "ignore":
            continue

        household_id = get_or_create_household_id(conn, person)
        sub_id = sub_id_by_name.get(product.lower())
        note = f"Import Einnahmen: {product}"

        mapped = map_months(months)
        if mapped:
            cycle, frequency = mapped
            existing = conn.execute(
                """
                SELECT id
                  FROM person_income_recurring
                 WHERE user_id = ?
                   AND household_id = ?
                   AND amount = ?
                   AND cycle = ?
                   AND frequency = ?
                   AND ifnull(subscription_id, 0) = ifnull(?, 0)
                 LIMIT 1
                """,
                (USER_ID, household_id, amount, cycle, frequency, sub_id),
            ).fetchone()

            if existing:
                conn.execute(
                    """
                    UPDATE person_income_recurring
                       SET note = ?, active = 1, start_date = ?
                     WHERE id = ? AND user_id = ?
                    """,
                    (note, TODAY, existing["id"], USER_ID),
                )
            else:
                conn.execute(
                    """
                    INSERT INTO person_income_recurring
                        (user_id, household_id, amount, currency_id, cycle, frequency, start_date, end_date, subscription_id, note, active)
                    VALUES
                        (?, ?, ?, ?, ?, ?, ?, NULL, ?, ?, 1)
                    """,
                    (USER_ID, household_id, amount, MAIN_CURRENCY_ID, cycle, frequency, TODAY, sub_id, note),
                )
            stats["income_recurring_upserted"] += 1
        else:
            conn.execute(
                """
                INSERT INTO person_income_entries
                    (user_id, household_id, amount, currency_id, income_date, subscription_id, note)
                VALUES
                    (?, ?, ?, ?, ?, ?, ?)
                """,
                (USER_ID, household_id, amount, MAIN_CURRENCY_ID, TODAY, sub_id, f"{note} ({months} Monate)"),
            )
            stats["income_entries_inserted"] += 1

    after_households = conn.execute(
        "SELECT COUNT(*) FROM household WHERE user_id = ?",
        (USER_ID,),
    ).fetchone()[0]

    conn.commit()
    conn.close()

    print("Import complete.")
    print(f"Household created: {max(0, after_households - before_households)}")
    print(f"Subscriptions upserted: {stats['subscriptions_upserted']}")
    print(f"Recurring incomes upserted: {stats['income_recurring_upserted']}")
    print(f"One-time incomes inserted: {stats['income_entries_inserted']}")


if __name__ == "__main__":
    main()

